### PR TITLE
ROR Integration - Live Import

### PR DESCRIPTION
**Pull request by vins01-4science**
**Please note that this development has been funded by the University of California - California Digital Library.**

## References
* Related to DSpace/dspace-angular#2719

## Description
This PR introduce a new `Organization` Data Provider for the live-import that retrieves entries using the ROR API .

## Instructions for Reviewers
Enable the live-import and then just try to import a new `Organization` entry with the ROR provider

List of changes in this PR:
* Configured a new data-provider for the live import
* Configured new mappings to retrieve items
* Created new metadata for the ROR related data
* Created new configuration for ROR beans 

## Checklist

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new libraries/dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies REST API endpoints, I've opened a separate [REST Contract](https://github.com/DSpace/RestContract/blob/main/README.md) PR related to this change.
- [x] If my PR includes new configurations, I've provided basic technical documentation in the PR itself.
- [x] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

### CODE REVIEW

File: dspace-api/src/main/java/org/dspace/importer/external/ror/service/RorImportMetadataSourceServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+            return jsonNode.at("/number_of_results").asInt();
+        } catch (URISyntaxException e) {
+            e.printStackTrace();
+        }
+        return 0;
+    }
+
+    private List<ImportRecord> searchById(String id) {
+
+        List<ImportRecord> adsResults = new ArrayList<>();
```
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/importer/external/ror/service/RorImportMetadataSourceServiceImpl.java:200]: The variable name is odd here, as this appears to refer to ADS, which is a different external service.  Small thing, but maybe change this to just be "importResults" or "results"?

File: dspace-api/src/main/java/org/dspace/importer/external/ror/service/RorImportMetadataSourceServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+            adsResults.add(transformSourceRecords(jsonNode.toString()));
+
+        } catch (URISyntaxException e) {
+            e.printStackTrace();
+        }
+        return adsResults;
+    }
+
+    private List<ImportRecord> search(String query) {
+        List<ImportRecord> adsResults = new ArrayList<>();
```
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/importer/external/ror/service/RorImportMetadataSourceServiceImpl.java:224]: Same here, rename to be "importResults" or something more generic?

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/RorImportMetadataSourceServiceIT.java [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+            assertThat(record.getValueList(), hasSize(11));
+
+            assertThat(
+                record.getSingleValue("organization.legalName"),
+                is("The University of Texas")
+            );
+            assertThat(record.getSingleValue("organization.identifier.ror"), is("https://ror.org/02f6dcw23"));
+            assertThat(record.getSingleValue("oairecerif.acronym"), is("UTHSCSA"));
+            assertThat(record.getSingleValue("oairecerif.identifier.url"), is("http://www.uthscsa.edu/"));
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/RorImportMetadataSourceServiceIT.java:69]: The "oairecerif" field checks here can be removed. These fields shouldn't be added to DSpace.

File: dspace-api/src/main/java/org/dspace/importer/external/ror/service/RorImportMetadataSourceServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+
+        @Override
+        public List<ImportRecord> call() throws Exception {
+            return searchById(query.getParameterAsClass("id", String.class));
+        }
+    }
+
+    /**
+     * This class is a Callable implementation to count the number of entries for an
+     * ADS query. This Callable use as query value to ADS the string queryString
```
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/importer/external/ror/service/RorImportMetadataSourceServiceImpl.java:155]: These Javadocs also reference ADS.  It looks like you may have reused some of the ADS code in general (which is fine). But, we should then remove any references to ADS please

**Review by tdonohue — CHANGES_REQUESTED** [review]
@vins01-4science : Thanks! Overall this looks good, but I've run into some small issues with getting it to work properly.  

I am able to search ROR via the Admin UI, but I cannot successfully import a new OrgUnit from ROR because I get this backend error:
```
2024-02-07 20:19:26,084 ERROR 023c6910-8f69-45fd-bfa1-5e2b69e0d5c2 ddf2a890-bec3-43e3-bdc3-5c6d6b605a28 org.dspace.app.rest.repository.handler.ExternalSourceEntryItemUriListHandler @ An error occured when trying to create item in collection with uuid: d8ab207f-6eda-4c42-b2e5-ef76c2586070
 java.sql.SQLException: bad_dublin_core schema=oairecerif.identifier.url. Metadata field does not exist!
   at org.dspace.content.DSpaceObjectServiceImpl.addMetadata(DSpaceObjectServiceImpl.java:223) ~[dspace-api-8.0-SNAPSHOT.jar:8.0-SNAPSHOT]
   at org.dspace.content.DSpaceObjectServiceImpl.addMetadata(DSpaceObjectServiceImpl.java:363) ~[dspace-api-8.0-SNAPSHOT.jar:8.0-SNAPSHOT]
   at org.dspace.external.service.impl.ExternalDataServiceImpl.createWorkspaceItemFromExternalDataObject(ExternalDataServiceImpl.java:99) ~[dspace-api-8.0-SNAPSHOT.jar:8.0-SNAPSHOT]
```

As you can see from the above error, the problem appears to be that you are using some invalid `oairecerif.*` metadata fields in the code of this PR.  Those need to be removed as they are specific to DSpace-CRIS at this time.  DSpace doesn't have an `oairecerif` metadata schema.  I've noted this again inline below.

Overall, the code looks good & I expect this will work once the `oairecerif.*` fields are removed.  I do have a number of other inline comments below, but most should be very easy to cleanup/fix.

(NOTE: I've also noticed when using this feature that a number of i18n keys are missing from the UI for this feature.  I'll add that feedback to https://github.com/DSpace/dspace-angular/pull/2719 so they can be fixed there).

**Review by paulo-graca — DISMISSED** [review]
It seems like this PR shares code with #9237 . I've just left my comments there. Basically, there are some missing java docs and schema usage suggestions.

**Comment by vins01-4science** [comment]
@tdonohue 
> @vins01-4science : Thanks! Overall this looks good, but I've run into some small issues with getting it to work properly.
> 
> I am able to search ROR via the Admin UI, but I cannot successfully import a new OrgUnit from ROR because I get this backend error:
> 
> ```
> 2024-02-07 20:19:26,084 ERROR 023c6910-8f69-45fd-bfa1-5e2b69e0d5c2 ddf2a890-bec3-43e3-bdc3-5c6d6b605a28 org.dspace.app.rest.repository.handler.ExternalSourceEntryItemUriListHandler @ An error occured when trying to create item in collection with uuid: d8ab207f-6eda-4c42-b2e5-ef76c2586070
>  java.sql.SQLException: bad_dublin_core schema=oairecerif.identifier.url. Metadata field does not exist!
>    at org.dspace.content.DSpaceObjectServiceImpl.addMetadata(DSpaceObjectServiceImpl.java:223) ~[dspace-api-8.0-SNAPSHOT.jar:8.0-SNAPSHOT]
>    at org.dspace.content.DSpaceObjectServiceImpl.addMetadata(DSpaceObjectServiceImpl.java:363) ~[dspace-api-8.0-SNAPSHOT.jar:8.0-SNAPSHOT]
>    at org.dspace.external.service.impl.ExternalDataServiceImpl.createWorkspaceItemFromExternalDataObject(ExternalDataServiceImpl.java:99) ~[dspace-api-8.0-SNAPSHOT.jar:8.0-SNAPSHOT]
> ```
> 
> As you can see from the above error, the problem appears to be that you are using some invalid `oairecerif.*` metadata fields in the code of this PR. Those need to be removed as they are specific to DSpace-CRIS at this time. DSpace doesn't have an `oairecerif` metadata schema. I've noted this again inline below.
> 
> Overall, the code looks good & I expect this will work once the `oairecerif.*` fields are removed. I do have a number of other inline comments below, but most should be very easy to cleanup/fix.
> 
> (NOTE: I've also noticed when using this feature that a number of i18n keys are missing from the UI for this feature. I'll add that feedback to [DSpace/dspace-angular#2719](https://github.com/DSpace/dspace-angular/pull/2719) so they can be fixed there).

@paulo-graca 

> It seems like this PR shares code with #9237 . I've just left my comments there. Basically, there are some missing java docs and schema usage suggestions.

Thank you both for the review, all the changes that you've requested have been addressed.
Let me know if you find something else!

**Review by tdonohue — APPROVED** [review]
👍 Thanks @vins01-4science ! I've re-reviewed and retested this today with the frontend PR.  Everything is working well.  I can successfully import an organization via ROR and when doing so I see the "ROR" logo next to the identifier.

I've also verified all my & @paulo-graca 's  prior feedback about the code has been addressed.  So, this looks good to merge

**Pull request merged**